### PR TITLE
kvutils & ledger-on-sql: Avoid a race condition in dispatching.

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
This changes the API of kvutils to allow for passing data out of the transaction, which makes it much easier to ensure the new head makes it to the dispatcher. Previously, we would use an `AtomicLong` to communicate the data, but this was problematic because values could arrive out in the wrong order. For example:

  - log index 5 is committed
  - `head` is updated to 6
  - the dispatcher is signalled with a head of 6
  - log index 6 is committed
  - log index 7 is committed
  - `head` is updated to 8
  - `head` is updated to 7
  - the dispatcher is signalled with a head of 7

In this scenario, we would have to wait until a new commit comes in before the indexer finds out about log index 7.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
Since this is a bug fix it could make sense to add an entry to the changelog.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/ValidationFailed.scala [hunk 1]
```diff
@@ -3,18 +3,12 @@
 
 package com.daml.ledger.validator
 
-sealed trait ValidationResult
+sealed trait ValidationFailed extends RuntimeException
```
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/ValidationFailed.scala]: Would it make sense to extends `NoStackTrace` here?

Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/ValidationFailed.scala]: Yeah, probably; we always handle it and honestly, I think we're going to get rid of it eventually.

**Comment by SamirTalwar-DA** [comment]
@stefanobaghino-da: Regarding the changelog, I'm skipping it because this stuff isn't "released" yet, so it'd just be confusing.

**Pull request merged**